libs/wingc/src/lsp/ast_traversal.rs

```diff
@@ -0,0 +1,887 @@
+use lsp_types::Position;
+
+use crate::ast::ExprKind::*;
+use crate::ast::StmtKind::*;
+use crate::ast::{Scope, *};
+
+/// While traversing the AST, this represent the context of a given node.
+#[derive(Debug, Clone, Copy)]
+pub struct TreeLocationContext<'a> {
+	pub scope: &'a Scope,
+	pub statement: &'a Stmt,
+	pub expression: Option<&'a Expr>,
+}
+
+/// Traverse the AST and find the symbol at the given position.
+pub fn find_symbol<'a>(scope: &'a Scope, position: &'a Position) -> Option<(TreeLocationContext<'a>, &'a Symbol)> {
+	let all_statements = get_statements_from_scope_no_context(scope);
+	for context in all_statements {
+		let statement = context.statement;
+		if !statement.span.contains(position) {
+			continue;
+		}
+
+		for symbol_lookup in get_symbols_from_statement(context).iter() {
+			let symbol = symbol_lookup.1;
+			if symbol.span.contains(position) {
+				return Some((symbol_lookup.0, symbol));
+			}
+		}
+
+		let all_expressions = get_expressions_from_statement(context);
+
+		for expression_context in all_expressions {
+			for symbol_lookup in get_symbols_from_expression(expression_context).iter() {
+				let symbol = symbol_lookup.1;
+				if symbol.span.contains(position) {
+					return Some((symbol_lookup.0, symbol));
+				}
+			}
+		}
+	}
+
+	None
+}
+
+/// Traverse the given statement context and return all expressions inside it (recursive).
+pub fn get_expressions_from_statement<'a>(context: TreeLocationContext) -> Vec<TreeLocationContext> {
+	let mut return_expressions: Vec<TreeLocationContext> = vec![];
+	match &context.statement.kind {
```

Comment:
Intuitively it feels like we should already have this type of traversal in the compiler code. Is there a way to reuse?

Reply:
Yeah this traversal code is similar to something we've written a few times. Unfortunately, I'm not sure the code I've written here is good enough for the rest of the compiler. Mostly because the compiler phases may need to have the iteration ordered in a certain way that I'm not sure these satisfy.

It's something I'd love to figure out soon though because I'm tired of writing this code.

https://github.com/winglang/wing/issues/1570